### PR TITLE
tune ML model to support custom model

### PR DESCRIPTION
**Pull request by ylwu-amzn**
Signed-off-by: Yaliang Wu <ylwu@amazon.com>

### Description
Reviewed in this PR https://github.com/opensearch-project/ml-commons/pull/452
Backport to 2.x branch.
 
### Issues Resolved
[List any issues this PR will resolve]
 
### Check List
- [x] New functionality includes testing.
  - [x] All tests pass
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [ ] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Review by Zhangxunmt — APPROVED** [review]

File: common/src/main/java/org/opensearch/ml/common/CommonValue.java [hunk 1]
```diff
@@ -44,6 +49,49 @@ public class CommonValue {
...
+            + MLModel.MODEL_STATE_FIELD
+            + "\" : {\"type\": \"keyword\"},\n"
+            + "      \""
+            + MLModel.MODEL_CONTENT_SIZE_IN_BYTES_FIELD
+            + "\" : {\"type\": \"long\"},\n"
+            + "      \""
+            + MLModel.MODEL_CONFIG_FIELD
+            + "\" : {\"properties\":{\""
+            + MODEL_TYPE_FIELD + "\":{\"type\":\"keyword\"},\""
+            + EMBEDDING_DIMENSION_FIELD + "\":{\"type\":\"long\"},\""
```
Comment (dhrubo-os) [inline common/src/main/java/org/opensearch/ml/common/CommonValue.java:77]: [nit]Does this field need to be `long`?

File: common/src/main/java/org/opensearch/ml/common/MLModel.java [hunk 2]
```diff
@@ -28,23 +34,71 @@ public class MLModel implements ToXContentObject {
     public static final String OLD_MODEL_CONTENT_FIELD = "content";
     public static final String MODEL_CONTENT_FIELD = "model_content";
 
+    public static final String DESCRIPTION_FIELD = "description";
+    public static final String MODEL_FORMAT_FIELD = "model_format";
+    public static final String MODEL_STATE_FIELD = "model_state";
+    public static final String MODEL_CONTENT_SIZE_IN_BYTES_FIELD = "model_content_size_in_bytes";
+    public static final String MODEL_CONTENT_HASH_FIELD = "model_content_hash";
```
Comment (dhrubo-os) [inline common/src/main/java/org/opensearch/ml/common/MLModel.java:41]: This field isn't self explanatory to me.

File: common/src/main/java/org/opensearch/ml/common/MLModel.java [hunk 3]
```diff
@@ -28,23 +34,71 @@ public class MLModel implements ToXContentObject {
...
+    private Instant lastUnloadedTime;
+
+    @Setter
+    private String modelId; // model chunk doc only
+    private Integer chunkNumber; // model chunk doc only
+    private Integer totalChunks; // model chunk doc only
+
     @Builder
-    public MLModel(String name, FunctionName algorithm, Integer version, String content, User user) {
+    public MLModel(String name, FunctionName algorithm, Integer version, String content, User user, String description, MLModelFormat modelFormat, MLModelState modelState, Long modelContentSizeInBytes, String modelContentHash, MLModelConfig modelConfig, Instant createdTime, Instant lastUploadedTime, Instant lastLoadedTime, Instant lastUnloadedTime, String modelId, Integer chunkNumber, Integer totalChunks) {
```
Comment (dhrubo-os) [inline common/src/main/java/org/opensearch/ml/common/MLModel.java:76]: `version` -> what if we want to have a version like 2.1?

File: common/src/main/java/org/opensearch/ml/common/model/MLModelConfig.java [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+import org.opensearch.common.xcontent.ToXContentObject;
+
+import java.io.IOException;
+
+@Setter
+@Getter
+public abstract class MLModelConfig implements ToXContentObject, NamedWriteable {
+
+    public static final String MODEL_TYPE_FIELD = "model_type";
+    public static final String ALL_CONFIG_FIELD = "all_config";
```
Comment (dhrubo-os) [inline common/src/main/java/org/opensearch/ml/common/model/MLModelConfig.java:22]: What is this field for?

File: common/src/main/java/org/opensearch/ml/common/model/TextEmbeddingModelConfig.java [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+    );
+
+    public static final String EMBEDDING_DIMENSION_FIELD = "embedding_dimension";
+    public static final String FRAMEWORK_TYPE_FIELD = "framework_type";
+
+    private Integer embeddingDimension;
+    private FrameworkType frameworkType;
+
+    @Builder
+    public TextEmbeddingModelConfig(String modelType, Integer embeddingDimension, FrameworkType frameworkType, String allConfig) {
```
Comment (dhrubo-os) [inline common/src/main/java/org/opensearch/ml/common/model/TextEmbeddingModelConfig.java:41]: can't we use @NonNull?

File: common/src/main/java/org/opensearch/ml/common/model/MLModelState.java [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+    LOADING,
+    LOADED,
+    PARTIALLY_LOADED,
+    UNLOADED;
+
+    public static MLModelState from(String value) {
+        try {
+            return MLModelState.valueOf(value);
+        } catch (Exception e) {
+            throw new IllegalArgumentException("Wrong model format");
```
Comment (dhrubo-os) [inline common/src/main/java/org/opensearch/ml/common/model/MLModelState.java:20]: Need to change the exception message.

**Review by dhrubo-os — APPROVED** [review]

File: common/src/main/java/org/opensearch/ml/common/MLModel.java (hunk 3 above)
Comment (ylwu-amzn) [reply common/src/main/java/org/opensearch/ml/common/MLModel.java:76]: Good point. The `version` part was int from initial release. I plan to revise that part later to not change this backport PR. I will create a Github issue to track this.

File: common/src/main/java/org/opensearch/ml/common/model/TextEmbeddingModelConfig.java (hunk 5 above)
Comment (ylwu-amzn) [reply common/src/main/java/org/opensearch/ml/common/model/TextEmbeddingModelConfig.java:41]: That's possible. I think that's not bad to write few more lines of code to make the error message more readable.

File: common/src/main/java/org/opensearch/ml/common/model/MLModelState.java (hunk 6 above)
Comment (ylwu-amzn) [reply common/src/main/java/org/opensearch/ml/common/model/MLModelState.java:20]: Good catch. Will fix this in next PR.

File: common/src/main/java/org/opensearch/ml/common/MLModel.java (hunk 2 above)
Comment (ylwu-amzn) [reply common/src/main/java/org/opensearch/ml/common/MLModel.java:41]: Do you have any suggestion? Basically this field is to track the model file's SHA256 hash value.

File: common/src/main/java/org/opensearch/ml/common/CommonValue.java (hunk 1 above)
Comment (ylwu-amzn) [reply common/src/main/java/org/opensearch/ml/common/CommonValue.java:77]: Good catch, that part should be int. Will keep this backport PR as is, and change in next PR.

File: common/src/main/java/org/opensearch/ml/common/MLModel.java (hunk 3 above)
Comment (ylwu-amzn) [reply common/src/main/java/org/opensearch/ml/common/MLModel.java:76]: https://github.com/opensearch-project/ml-commons/issues/472

**Pull request merged**